File: absolute_min_max.py
MAX = 250001


class BinaryIndexedTree:

    BITree = []

    def __init__(self, size):
        self.size = size + 2
        self.BITree = [0] * self.size

    def query(self, index):
        index += 1
        ans = 0

        while(index >= 1):
            ans += self.BITree[index]
            index -= index & (-index)

        return ans

    def update(self, index, value):
        index += 1

        while(index <= n):
            self.BITree[index] += value
            index += index & (-index)

    def LRquery(self, left, right):
        return self.query(right) - self.query(left-1)


def countOfSubarrays():
    li = []

    for _ in range(n):
        li.append([])

    tree_obj = BinaryIndexedTree(n+2)

    stack = []

    for i in range(n):
        while(len(stack) != 0 and arr[stack[-1]] <= arr[i]):
            stack.pop()

        if(len(stack) == 0):
            tree_obj.update(i, 1)

        else:
            li[stack[-1]].append(i)

        stack.append(i)

    stack = []
    next_greatest = [0]*MAX

    for i in range(n-1, -1, -1):
        while(len(stack) != 0 and arr[stack[-1]] >= arr[i]):
            stack.pop()

        if(len(stack) == 0):
            next_greatest[i] = n

        else:
            next_greatest[i] = stack[-1]

        stack.append(i)

    valid_subarrays = 0

    for i in range(n):
        for j in li[i]:
            tree_obj.update(j, 1)

        valid_subarrays += tree_obj.LRquery(i, next_greatest[i]-1)
        tree_obj.update(i, -1)

    return valid_subarrays


t = int(input())

for i in range(t):
    n = int(input())
    arr = list(map(int, input().split()))

    answer = 0
   
    answer += countOfSubarrays()
    arr.reverse()
    answer += countOfSubarrays()

    count_duplicate = 1

    for i in range(1, n):
        if(arr[i] != arr[i-1]):
            answer -= (count_duplicate * (count_duplicate+1))//2
            count_duplicate = 1

        else:
            count_duplicate += 1

    answer -= (count_duplicate * (count_duplicate+1))//2

    print(answer)

File: absolute_min_max.cpp
#include <bits/stdc++.h>
#define ll long long
#define MAX 250002

using namespace std;

ll n;
ll arr[MAX];

struct binaryIndexedTree
{

    ll size;
    ll BITree[MAX];

    void init(ll sz)
    {
        size = sz + 2;
        BITree[MAX] = {0};
    }

    void update(ll idx, ll value)
    {
        for (idx = idx + 1; idx <= size; idx += (idx & -idx))
        {
            BITree[idx] += value;
        }
    }

    ll query(ll idx)
    {
        ll query_result = 0;
        for (idx = idx + 1; idx >= 1; idx -= (idx & -idx))
        {
            query_result += BITree[idx];
        }
        return query_result;
    }

    ll LRquery(ll left, ll right)
    {
        return query(right) - query(left - 1);
    }
};

ll countOfSubarrays()
{

    vector<ll> li[MAX];
    vector<ll> stack;
    binaryIndexedTree tree_object;

    for (ll i = 0; i < n; i++)
    {
        li[i].clear();
    }

    tree_object.init(n);

    for (ll i = 0; i < n; i++)
    {
        while (!stack.empty() && arr[stack.back()] <= arr[i])
        {
            stack.pop_back();
        }

        if (stack.empty())
            tree_object.update(i, 1);

        else
            li[stack.back()].push_back(i);

        stack.push_back(i);
    }

    stack.clear();

    ll next_greatest[MAX] = {0};

    for (ll i = n - 1; i >= 0; i--)
    {
        while (!stack.empty() && arr[stack.back()] >= arr[i])
        {
            stack.pop_back();
        }

        if (stack.empty())
            next_greatest[i] = n;

        else
            next_greatest[i] = stack.back();

        stack.push_back(i);
    }

    ll valid_subarrays = 0;

    for (ll i = 0; i < n; i++)
    {
        for (ll j : li[i])
            tree_object.update(j, 1);

        valid_subarrays += tree_object.LRquery(i, next_greatest[i] - 1);
        tree_object.update(i, -1);
    }

    return valid_subarrays;
}

int main()
{
    ll t;
    cin >> t;

    while (t--)
    {
        cin >> n;

        for (ll i = 0; i < n; i++)
            cin >> arr[i];

        ll answer = 0;
        answer += countOfSubarrays();
        reverse(arr, arr + n);
        answer += countOfSubarrays();

        ll duplicate_count = 1;

        for (ll i = 1; i < n; i++)
        {
            if (arr[i] != arr[i - 1])
            {
                answer -= (duplicate_count * (duplicate_count + 1)) / 2;
                duplicate_count = 1;
            }
            else
                duplicate_count++;
        }

        answer -= (duplicate_count * (duplicate_count + 1)) / 2;

        cout << answer << endl;
    }
    return 0;
}